### PR TITLE
[RL] Extend RLTrainer.setup() to accept an optional host_mesh parameter

### PR DESCRIPTION
**Pull request by daniellepintz**
Extend RLTrainer.setup() to accept an optional host_mesh parameter for multi-node mode. When provided, dedicates whole nodes to trainer/generator roles using HostMesh.slice() and sets CUDA_VISIBLE_DEVICES via Provisioner to isolate GPU allocations.

### CODE REVIEW

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 1]
```diff
@@ -160,12 +179,29 @@ def _compute_world_size(p: "ParallelismConfig") -> int:
...
+
+        Args:
+            host_mesh: Optional multi-node HostMesh. When provided,
+                whole nodes are dedicated to trainer vs generator
+                roles instead of partitioning GPUs on a single host.
+            trainer_nodes: Number of nodes for the trainer (required when
+                host_mesh is provided).
+            generator_nodes: Number of nodes for the generator (required when
+                host_mesh is provided).
+            gpus_per_node: GPUs per node (required when host_mesh is provided).
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py]: this is assuming no heterogeneity is allowed? please explicit comment this

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 2]
```diff
@@ -182,19 +218,88 @@ async def setup(self):
...
+
+            # Ensure TP isn't split across nodes (requires fast NVLink)
+            trainer_tp = config.trainer.parallelism.tensor_parallel_degree
+            generator_tp = config.generator.parallelism.tensor_parallel_degree
+            if trainer_gpus_per_node < trainer_tp:
+                raise ValueError(
+                    f"trainer tensor_parallel_degree ({trainer_tp}) exceeds "
+                    f"trainer_gpus_per_node ({trainer_gpus_per_node}). "
+                    f"Use fewer trainer_nodes to keep TP on a single node."
+                )
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py]: why do we care if it's user decision?

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 3]
```diff
@@ -182,19 +218,88 @@ async def setup(self):
...
+            trainer_host_mesh = host_mesh.slice(hosts=slice(0, trainer_nodes))
+            generator_host_mesh = host_mesh.slice(
+                hosts=slice(trainer_nodes, trainer_nodes + generator_nodes)
+            )
+
+            # Use Provisioner to set CUDA_VISIBLE_DEVICES so each role
+            # only sees its own GPUs and doesn't conflict with other
+            # processes on the node
+            trainer_provisioner = Provisioner(total_gpus=gpus_per_node)
+            generator_provisioner = Provisioner(total_gpus=gpus_per_node)
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py:269]: what's the rationale behind this vs. one Provisioner allocate resource for every participant?

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 4]
```diff
@@ -160,12 +179,29 @@ def _compute_world_size(p: "ParallelismConfig") -> int:
...
         )
 
-    async def setup(self):
+    async def setup(
+        self,
+        *,
+        host_mesh=None,
+        trainer_nodes: int | None = None,
+        generator_nodes: int | None = None,
+        gpus_per_node: int | None = None,
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py:194]: where is this used

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 5]
```diff
@@ -160,12 +179,29 @@ def _compute_world_size(p: "ParallelismConfig") -> int:
             * p.context_parallel_degree
         )
 
-    async def setup(self):
+    async def setup(
+        self,
+        *,
+        host_mesh=None,
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py:191]: didn't see this used / tested anywhere

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 6]
```diff
@@ -160,12 +179,29 @@ def _compute_world_size(p: "ParallelismConfig") -> int:
...
+    ):
         """Spawn Monarch actors on separate meshes and initialize weights.
 
         Creates separate GPU meshes for trainer and generator, a CPU mesh for
         the grader, and synchronizes initial weights from trainer to generator.
         Must be called before :meth:`train`.
+
+        Args:
+            host_mesh: Optional multi-node HostMesh. When provided,
+                whole nodes are dedicated to trainer vs generator
```
Comment (acisseJZhong) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py:204]: does this support multi-hosts but still partition gpus across trainer/generator? For example 2 hosts, with 4 generator gpus, 12 ttrainer gpus.

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 7]
```diff
@@ -147,6 +147,25 @@ def __init__(self, config: Config):
...
+                await mesh.stop()
+            except Exception:
+                pass
+        self._proc_meshes = []
+
+    def _extract_result(self, result, has_gpus: bool = True):
+        """Extract rank 0 result, handling both single and multi-node meshes."""
+        kwargs = {}
+        if self._multi_node:
+            kwargs["hosts"] = 0
```
Comment (acisseJZhong) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py:171]: qq: what will this do?

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 8]
```diff
@@ -321,14 +426,14 @@ async def train(self):
...
-                .get()
-                .item(gpus=0)
+            episodes = self._extract_result(
+                self.generator.generate.call(train_prompts, train_answers).get()
             )
 
             # 2. Grader computes rewards per episode
-            episodes = self.grader.score.call(episodes).get().item()
+            episodes = self._extract_result(
+                self.grader.score.call(episodes).get(), has_gpus=False
```
Comment (acisseJZhong) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py:429]: qq: does this mean reward calculation is purely on cpu?

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 6 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:204]: it currently does not - currently it is one actor per host. We can consider changing this later

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 8 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:429]: yes exactly

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 7 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:171]: Monarch actor endpoints return results from all ranks in the mesh. This picks out rank 0's result by indexing into the host and GPU dimensions as needed (multi-node meshes have an extra host dimension). This should be used in cases where all ranks return the same result.

(Added to docstring)

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 5 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:191]: This is used in the internal-only code in the linked diff. We could probably add a CI test that uses this, I can follow up with that if you think it makes sense? will require multiple nodes

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 4 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:194]: This is used in the internal-only code in the linked diff

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 1 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py]: yes, added comment

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 2 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py]: I wanted to capture cases where TP is *unnecessarily* split across nodes, because of the calculation above:
```
trainer_gpus_per_node = self.trainer_world_size // trainer_nodes
```
Updated the check here to capture that, thanks

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 3 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:269]: IMO it's clearer to read if the Provisioner just handles GPUs for one host; if we wanted to use the same Provisioner and do `allocate` calls for different hosts we would need to have some sort of `reset` in between the allocate calls. cc @joecummings in case you have any thoughts here

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 9]
```diff
@@ -147,6 +147,31 @@ def __init__(self, config: Config):
...
+    async def cleanup(self):
+        """Stop all proc meshes to release GPU memory."""
+        for mesh in self._proc_meshes:
+            try:
+                await mesh.stop()
+            except Exception:
+                pass
+        self._proc_meshes = []
+
+    def _extract_result(self, result, has_gpus: bool = True):
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py]: let's be more explicit here
```suggestion
    def _get_rank_0_value(self, result, has_gpus: bool = True):
```

Comment (tianyu-l) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py]: what's the type of `result`? I can only see that you can call `.item()` on it, is it always a tensor?

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:269]: It's not clear to me how it can work when either trainer / generator spans multiple node.

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 10]
```diff
@@ -182,19 +226,90 @@ async def setup(self):
...
+            # Compute GPUs per node for each role based on the config's
+            # world size and number of nodes allocated to that role
+            trainer_gpus_per_node = self.trainer_world_size // trainer_nodes
+            generator_gpus_per_node = self.generator_world_size // generator_nodes
+
+            if trainer_gpus_per_node * trainer_nodes != self.trainer_world_size:
+                raise ValueError(
+                    f"trainer_world_size ({self.trainer_world_size}) must be "
+                    f"evenly divisible by trainer_nodes ({trainer_nodes})"
+                )
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py]: I think these checks should happen in it's own component, e.g. provisioner.

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 11]
```diff
@@ -182,19 +226,90 @@ async def setup(self):
...
+            # Ensure TP isn't unnecessarily split across nodes
+            trainer_tp = config.trainer.parallelism.tensor_parallel_degree
+            generator_tp = config.generator.parallelism.tensor_parallel_degree
+            if trainer_gpus_per_node < trainer_tp and gpus_per_node >= trainer_tp:
+                raise ValueError(
+                    f"trainer tensor_parallel_degree ({trainer_tp}) exceeds "
+                    f"trainer_gpus_per_node ({trainer_gpus_per_node}) but fits "
+                    f"on a single node ({gpus_per_node} GPUs). "
+                    f"Use fewer trainer_nodes to keep TP on a single node."
+                )
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py]: > I wanted to capture cases where TP is unnecessarily split across nodes, because of the calculation above:

I don't think we need to do this for users, we are not building auto parallel, so users should be responsible for their own parallel configs.

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 9 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py]: `result` is a Monarch [ValueMesh](https://github.com/meta-pytorch/monarch/blob/1000989be29e4184aee05abb4164a8d48a6c3fa0/python/monarch/_src/actor/actor_mesh.py#L738) — the return value of `.get()` on a Monarch actor call. It's a multi-dimensional container indexed by mesh dimensions (e.g., hosts, gpus), holding the return value from each rank. 

The [`.item(**kwargs)`](https://github.com/meta-pytorch/monarch/blob/1000989be29e4184aee05abb4164a8d48a6c3fa0/python/monarch/_src/actor/actor_mesh.py#L811) call extracts a single value by indexing into those dimensions.

The types in the container, and the return type of `.item()` will be whatever type the Monarch endpoint returns.

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 11 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py]: Ok, sg will remove

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 3 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:269]: If trainer/generator spans multiple nodes, the provisioner will allocate the appropriate GPUs on each node (setting CUDA_VISIBLE_DEVICES); the bootstrap function (allocate) is called _independently_ on each node when we call `spawn_procs`, so this works well.

but if we tried to use the same provisioner for the generator spawn_procs as well, the provisioner would already have the "state" from the trainer spawn_procs, e.g. `next_gpu` would already have a value that is not 0.

It is a bit confusing, pls lmk if you have more questions

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 10 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py]: I am not sure I agree with this, because then the provisioner would need to know about the trainer / generator world size; right now all the provisioner knows about is managing the number of GPUs on one node and knows which are in use and which are free

I guess if we expanded the scope of the Provisioner this could work - how do you envision it?

I do think these checks should be converted to asserts that happen before `trainer_gpus_per_node` is calculated

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 3 above)
Comment (joecummings) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:269]: The rationale makes sense - the root of the problem as pointed out by @allenwang28 is that although Monarch does internally keep track of host->gpu offsets, it doesn't expose this API (yet - we should ask them to). 

However, I have a preference for keeping the training loop code cleaner. To that end, we could temporarily hold these offsets in the Provisioner itself rather.

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 12]
```diff
@@ -147,6 +147,31 @@ def __init__(self, config: Config):
...
+    async def cleanup(self):
+        """Stop all proc meshes to release GPU memory."""
+        for mesh in self._proc_meshes:
+            try:
+                await mesh.stop()
+            except Exception:
+                pass
+        self._proc_meshes = []
+
+    def _get_rank_0_value(self, result, has_gpus: bool = True):
```
Comment (joecummings) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py:161]: I'm not sure I understand why this is needed

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 3 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:269]: > we could temporarily hold these offsets in the Provisioner itself rather.

could you explain what you mean exactly? Also, can this be done as a follow up?

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py [hunk 13]
```diff
@@ -182,19 +226,74 @@ async def setup(self):
...
+                trainer_nodes is None
+                or generator_nodes is None
+                or gpus_per_node is None
+            ):
+                raise ValueError(
+                    "trainer_nodes, generator_nodes, and gpus_per_node are "
+                    "required when host_mesh is provided"
+                )
+            # Validate that world sizes are evenly divisible by node counts
+            assert self.trainer_world_size % trainer_nodes == 0, (
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/simple_grpo_sum_digits.py:243]: Since we don't allow trainer and generator sharing a node, does it make sense to require each actor to occupy all the gpus on the same node? Like 
```
assert self.trainer_world_size == trainer_nodes * gpus_per_node
```

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 12 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:161]: In many cases we need to get the result of a Monarch endpoint from the controller, and the `get()` call returns a Monarch multi-dimensional container, that we need to index into with `.item(gpus=<rank>)` to get the actual result we want. For example, getting the generated episodes:

https://github.com/pytorch/torchtitan/blob/18a46d427e2e6abf5c84073d349d44fd3c520d81/torchtitan/experiments/rl/simple_grpo_sum_digits.py#L324-L328

 We use rank 0 for clarity, since all ranks have the same result, and we are calling this from the controller.

As you can see we have this in existing training loop, in this PR I am just moving it into a separate method

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 13 above)
Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:243]: Good question, I definitely see your point.

My only concern is then the configs for smaller models like `rl_grpo_qwen3_0_6b` wouldn't work out of the box on multi-node, you would have to manually change the parallelism settings and getting an AssertionError might be annoying.

I would lean towards not enforcing that for now, so we can use existing configs out of the box, but happy to keep discussing this after I land this PR; can always add later.

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 3 above)
Comment (joecummings) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:269]: Sure this can be a follow up

File: torchtitan/experiments/rl/simple_grpo_sum_digits.py (hunk 12 above)
Comment (joecummings) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:161]: This seems to add extra code just to do the item() call, which is more clear IMO - why do we need it?

Comment (daniellepintz) [reply torchtitan/experiments/rl/simple_grpo_sum_digits.py:161]: yes, the thing is for multi node the "container" of results becomes more complex / multi dimensional, and we need this code to work for the multi node **and** the single code case
